### PR TITLE
fix: use fasttimers for all connection timeouts

### LINKED ISSUE #3410: `UND_ERR_CONNECT_TIMEOUT` errors thrown when there is CPU intensive code on the event loop
## Bug Description

Work on the event loop can interrupt the Undici lifecycle for making requests, causing errors to be thrown even when there is no problem with the underlying connection. For example, if a fetch request is started and then work on the event loop takes more than 10 seconds (default connect timeout), Undici will throw a `UND_ERR_CONNECT_TIMEOUT` error even if the connection could be established very quickly.

I believe what is happening is:

1. When the fetch request is started, Undici starts the work to make a connection. Undici calls `setTimeout` with the value of the `connectTimeoutMs` to throw an error and cancel the connection if it takes too long (https://github.com/nodejs/undici/blob/main/lib/core/connect.js). It makes a call to `GetAddrInfoReqWrap` (https://github.com/nodejs/node/blob/main/lib/dns.js#L221), but this is asynchronous and processing of the callback will be delayed until the next event loop.
2. User tasks block the event loop for a long period of time.
3. The `onConnectTimeout` timer is run because the previous task took longer than the timeout. `onConnectTimeout` calls `setImmediate` with a function to destroy the socket and throw the error. https://github.com/nodejs/undici/blob/main/lib/core/connect.js
4. The `GetAddrInfoReq` lookup callback (`emitLookup` in `node:net`) is run. This code begins the TCP connection (`internalConnect` is called in https://github.com/nodejs/node/blob/main/lib/net.js#L1032) but that is also asynchronous, so it won't finish in this round of the event loop.
5. The `setImmediate` function is run in the next phase which destroys the socket and throws the `UND_ERR_CONNECT_TIMEOUT` error.
6. Undici never gets a chance to handle the TCP connection response.

Internally at Vercel, we have been seeing a high number of these `UND_ERR_CONNECT_TIMEOUT` issues while pre-rendering pages in our Next.js application. I can't run this task on my local machine so it's harder to debug, but it's a CPU intensive task and moving fetch requests to a worker thread eliminated the Undici errors. We tried other suggestions (like `--dns-result-order=ipv4first` and verified that we were not seeing any packet loss) that did not resolve the issue. Increasing the connect timeout resolves the issue in the reproduction but _not_ the issue in our Next.js build (which I can't explain).

## Reproducible By

A minimal reproduction is available at https://github.com/mknichel/undici-connect-timeout-errors.

We can reproduce the behavior on Node 18.x and 20.x and with the `5.24.0` and the latest version of Undici (`6.19.2`)

## Expected Behavior

The Undici request lifecycle could operate on a separate thread that does not get blocked by user code. By separating it out from the user code, this would remove impact of any user code on requests.

To test this theory, we created a dispatcher that proxied the fetch request to a dedicated worker thread (`new Worker` from `worker_threads`). This eliminated all the Undici errors that we were seeing in our Next.js build.

## Logs & Screenshots

In the minimal reproduction, the error is:

```
TypeError: fetch failed
    at fetch (/Users/mknichel/code/tmp/undici-connect-timeout-errors/node_modules/.pnpm/undici@6.19.2/node_modules/undici/index.js:112:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at fetchExample (/Users/mknichel/code/tmp/undici-connect-timeout-errors/index.ts:21:20)
    at main (/Users/mknichel/code/tmp/undici-connect-timeout-errors/index.ts:66:3) {
  [cause]: ConnectTimeoutError: Connect Timeout Error
      at onConnectTimeout (/Users/mknichel/code/tmp/undici-connect-timeout-errors/node_modules/.pnpm/undici@6.19.2/node_modules/undici/lib/core/connect.js:190:24)
      at /Users/mknichel/code/tmp/undici-connect-timeout-errors/node_modules/.pnpm/undici@6.19.2/node_modules/undici/lib/core/connect.js:133:46
      at Immediate._onImmediate (/Users/mknichel/code/tmp/undici-connect-timeout-errors/node_modules/.pnpm/undici@6.19.2/node_modules/undici/lib/core/connect.js:174:9)
      at process.processImmediate (node:internal/timers:478:21) {
    code: 'UND_ERR_CONNECT_TIMEOUT'
```

In our Next.js builds, the error is:

```
TypeError: fetch failed
    at node:internal/deps/undici/undici:12618:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async s (elided path)
    at async elided path {
  cause: ConnectTimeoutError: Connect Timeout Error
      at onConnectTimeout (node:internal/deps/undici/undici:7760:28)
      at node:internal/deps/undici/undici:7716:50
      at Immediate._onImmediate (node:internal/deps/undici/undici:7748:13)
      at process.processImmediate (node:internal/timers:478:21)
      at process.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'UND_ERR_CONNECT_TIMEOUT'
    }
  }
```

## Environment

The reproduction repo was erroring for me on Mac OS 14.4, while internally we are seeing issues on AWS EC2 Intel machines.

### Additional context

Vercel/Next.js users have reported `UND_ERR_CONNECT_TIMEOUT` issues to us:

- https://github.com/vercel/vercel/issues/11692
- https://github.com/vercel/next.js/issues/66373

### LINKED ISSUE #3553: TIMEOUT_IDLE should include event loop lag
Tee `TIMEOUT_IDLE` state used by the keep alive timeout should include the event loop lag since it's tracking a remote servers time.

I think the recent fast timers event loop lag fixes has broken this. We should not use fast timers for this state.

### PR DESCRIPTION
**Pull request by Uzlopak**
Fixes #3410
Fixes #3553

currently WIP!

Assessment:

The PR #3495 was imho the right step to fix the event loop issues. But Connection issues were still reported. Maybe the issue still existing because timeouts smaller than 1000ms will be handled by native Timers. Maybe those have to handled by the low resolution fast timer logic too, so that event loop blockers are less effecting the processing of requests.

Maybe we should increase the resolution of the timers to 200ms?! It should be still more performant than having a gazillion of native timers? 

Maybe we should change this else if to an if and maybe check if the timer missed a tick (with the side effect that we need to call `performance.now()`).
https://github.com/nodejs/undici/blob/89a46dd54f7d7db7513c435fac8042769ee9e9b5/lib/util/timers.js#L159

So if we skipped a tick, we immediatly run the timer, because we are just too late?

But on the other hand, this could actually trigger timeouts too early?!

So yeah, I would prefer to increase the resolution.

### CODE REVIEW

**Comment by KhafraDev** [comment]
Are the fast timers even needed anymore after https://github.com/nodejs/node/pull/46579/?

**Comment by ronag** [comment]
Yea. I don't think that PR makes much difference.

**Comment by ronag** [comment]
> Are the fast timers even needed anymore after [nodejs/node#46579](https://github.com/nodejs/node/pull/46579)?

Fast timers also adjust for event loop lag.

**Comment by Uzlopak** [comment]
Ok, then i change this PR to rip out fast timers

**Comment by ronag** [comment]
No, don't remove fast timers.

**Review by ronag — CHANGES_REQUESTED** [review]
Do not remove fast timers.

**Comment by Uzlopak** [comment]
@ronag 

Oh! good that i had a backup :). reverted

File: lib/util/timers.js [hunk 1]
```diff
@@ -362,6 +362,31 @@ module.exports = {
...
+   * expires.
+   * @param {number} delay The time, in milliseconds that the timer should
+   * wait before the specified function or code is executed.
+   * @param {*} [arg] An optional argument to be passed to the callback function
+   * when the timer expires.
+   * @returns {FastTimer}
+   */
+  setFastTimeout (callback, delay, arg) {
+    // If the delay is less than or equal to the RESOLUTION_MS value return a
+    // native Node.js Timer instance.
```
Comment (Uzlopak) [inline lib/util/timers.js:378]:
```suggestion
```

**Review by ronag — APPROVED** [review]

**Comment by Uzlopak** [comment]
@ronag 

but one thing maybe for interest

The test run for ripped out fast timers passes the tests

https://github.com/nodejs/undici/actions/runs/10714949877

This indicates that event loop lags are also handled properly by native timers.

But I guess we want to keep our fast timers solution due to the fact, that it is still less resource hungry.

**Comment by ronag** [comment]
> This indicates that event loop lags are also handled properly by native timers.

This can't be true as it's based on the uv loop....

**Review by metcoder95 — APPROVED** [review]

**Comment by Uzlopak** [comment]
let me fix the request-timeout tests.

**Comment by Uzlopak** [comment]
@ronag 

Should we also use setFastTimeout here?

https://github.com/nodejs/undici/blob/61a4b699c9f851d8c3d5b2a6f51f8c477249b921/lib/core/connect.js#L169

**Comment by ronag** [comment]
Yes

**Comment by Uzlopak** [comment]
adapted so far, but not all tests are reactivated. so please not merge yet

File: lib/core/connect.js [hunk 2]
```diff
@@ -116,11 +120,14 @@ function buildConnector ({ allowH2, maxCachedSessions, socketPath, timeout, sess
         })
     } else {
       assert(!httpSocket, 'httpSocket can only be sent on TLS update')
+
+      port = port || 80
```
Comment (ronag) [inline lib/core/connect.js:124]: why?

File: lib/core/connect.js [hunk 3]
```diff
@@ -96,6 +98,8 @@ function buildConnector ({ allowH2, maxCachedSessions, socketPath, timeout, sess
 
       const session = customSession || sessionCache.get(sessionKey) || null
 
+      port = port || 443
```
Comment (ronag) [inline lib/core/connect.js:101]: why?

Comment (Uzlopak) [reply lib/core/connect.js:101]: I want to see where the socket error is originating.

**Comment by Uzlopak** [comment]
@ronag 

Because you asked, why i change the port. The thing is, that I want to use the port for the connection timeout to have in the logs the chance to determine atleast which server/connection resulted in a connection timeout.

**Review by ronag — APPROVED** [review]

**Comment by trivikr** [comment]
The backport fails to 6.x, as the refactor of fast timers in https://github.com/nodejs/undici/pull/3495 is not backported to 6.x

**Pull request merged**